### PR TITLE
Campaign skip feature

### PR DESCRIPTION
**Pull request by jaredstrong89**
I have implemented the campaign skip (story skip) functionality. When the "Skip Campaign" button is clicked on the client application (only visible when player has not cleared chapter 10), player data will be adjusted as described in the Dragalia Lost Wiki.

- Player level: 60
- Stamina: 999
- Getherwings: 99
- All quest first clear and endeavors completed on Easy mode from chapters 1-10, granting Wyrmite rewards
- All quest stories marked as "read" from chapters 1-10
- Story units and dragon rewards added to collection
- Tutorials skipped through chapter 10
- Story progression Halidom facilities added to storage with their appropriate levels

I also included a fairly detailed unit test, which will check player level, Halidom buildings, etc.

[Api Doc](https://dragalia-api-docs.readthedocs.io/en/latest/dragalia/story_skip_skip.html)
[Dragalia Lost Wiki](https://dragalialost.wiki/w/Beginner%27s_Guide#The_Story_Skip_Option)

This was my first dive into a C# web app, so hopefully my code fits with your coding conventions.

### CODE REVIEW

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/IStorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/IStorySkipService.cs:3]: I would get rid of this interface - I know most services in the codebase have an interface, but I've stopped writing them recently as they don't add much value if you only have one implementation.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:188]: Why do you need this `SaveChanges`? Calling it in a loop will be very expensive

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:98]: Would it be better to use `FortRepository.Builds` here, to skip the extra lookup, and to do the filtering as part of the DB query? This might work:
```suggestion
        List<DbFortBuild> buildListEnum = await fortRepository.Builds.Where(x => fortConfigs.Keys.Contains(x.PlantId)).ToListAsync();
        foreach (BuildList buildList in buildListEnum)
        {
            fortConfig = fortConfigs[build.PlantId];
            if (build.Level < fortConfig.Level)
```

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:129]: Do we have to open treasure as part of the story skip? The wiki doesn't mention it anywhere

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:108]: Does a bit more than increase level, maybe best to rename?
```suggestion
    public async Task UpdateUserData()
```

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:83]: Nitpick: you can destructure here to make this cleaner

```suggestion
        foreach ((FortPlants fortPlant, FortConfig fortConfig) in fortConfigs)
        {
```

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:44]: I don't think this does anything

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs:51]: It may be worth caching these in a `FrozenSet` as a private field in `StorySkipService` so that they don't have to  be recalculated for each request -- they won't change after all

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs:44]: Why do you need the filter on `Id > 100000000` and that the 6th digit is `1`?

**Review by SapiensAnatis — CHANGES_REQUESTED** [review]
Thank you for this! A very good first attempt for someone new to the codebase and C# as well. I haven't yet pulled this down to test it locally but just some initial feedback.

I think that we may want some specialist infrastructure here to avoid making hundreds of DB calls - yes story skip is only once per account and boots you to the title screen, but it will stress the DB if people do it quite frequently. I'm mainly concerned here about some of the 'get single entity' calls in the loops -- for instance `ReadStory` I think  will call `FindAsync` 104 times and `GetQuestDataAsync` 114 times.

I'll have a think about what that specialist infrastructure could look like - usually we just preload stuff manually and skip the otherwise async/query checks via a dict lookup, but maybe we could use a merge statement or similar. Hope this is at least something to start on.

**Comment by SapiensAnatis** [comment]
P.S. csharpier is an auto-formatter, to get it to pass run
```bash
dotnet tool restore
dotnet csharpier .
```
from the repository root. I'm also happy to run it before merging if you'd prefer that.

**Comment by jaredstrong89** [comment]
> P.S. csharpier is an auto-formatter, to get it to pass run
> 
> ```shell
> dotnet tool restore
> dotnet csharpier .
> ```
> 
> from the repository root. I'm also happy to run it before merging if you'd prefer that.

These tools are new to me, so this is good to know. Thank you!

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs:44]: While comparing the different quest Ids, I found that index 6 was a flag for the easy mode missions. I tested this out, and it worked nicely.
1 = easy, 2 = hard, 3 = very hard
The chapter number and quest number are encoded in there as well.

Because I was doing the substring, I was hitting a string length error on that Id. Instead of investigating further, I filtered for Id > 100000000.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs:51]: I will look into this.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:44]: Yes, I think you're right. I copy/pasted this from the mission completion service. I made just the essential changes, but I can clean it up.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:83]: This is good to know! I'll try it out.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:108]: Noted.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:129]: This was one of those nuances that I was unsure about. I can remove the treasure opening.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:98]: That does sound simpler... Noted.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:188]: Maybe you can help me understand database updates better. Should I only call the update service at the end after all data changes were made?

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/IStorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/IStorySkipService.cs:3]: Ah ok, that sounds great. Yeah most of the larger files had interfaces, so I followed suit.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipController.cs:86]: Ok, that sounds good.

**Comment by jaredstrong89** [comment]
> Thank you for this! A very good first attempt for someone new to the codebase and C# as well. I haven't yet pulled this down to test it locally but just some initial feedback.
> 
> I think that we may want some specialist infrastructure here to avoid making hundreds of DB calls - yes story skip is only once per account and boots you to the title screen, but it will stress the DB if people do it quite frequently. I'm mainly concerned here about some of the 'get single entity' calls in the loops -- for instance `ReadStory` I think will call `FindAsync` 104 times and `GetQuestDataAsync` will call `FindAsync` 114 times.
> 
> I'll have a think about what that specialist infrastructure could look like - usually we just preload stuff manually and skip the otherwise async/query checks via a dict lookup, but maybe we could use a merge statement or similar. Hope this is at least something to start on.

Thank you for all the feedback! You provided lots of ideas for me to think about. I have spent some of my free time in the past 2 weeks learning how this repo works, figuring out how to develop .NET on a Mac, with little C# experience and no Postgres knowledge. It's been a fun challenge.

When I got started writing the code, my strategy was to tie in to the other services to be thorough with granting rewards (like story characters and dragons), trying to not miss a step in each of the processes. But I see your point. We definitely want to be less chatty with the database.

Now that I can wrap my head around this thing much better, I have a better understanding of the database and how it relates to the client application. Going forward, I could pull the necessary data at one time, eliminating multiple lookups per table. I could remove references to those other services that handle updates, performing table updates/inserts directly. I will probably have to set up more lists, such as story character unlocks.

**Comment by jaredstrong89** [comment]
Lots of changes with this last commit. There should be way less database lookups. And I ran CSharpier this time!

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:80]: This will still get computed on every execution of this method. You need to hoist it as a `static readonly` variable in the class itself so that it is shared between instances:

```cs
public class StorySkipService
{
    private static readonly FrozenSet<QuestData> // ...
```

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (SapiensAnatis) [inline DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs:29]: Not sure if you realize, but every time you do a `Any()`, `Count()` or iterate through with `foreach` on an `IQueryable` you are executing a database query. I would suggest realizing the results of the query before you do your processing:

```suggestion
        List<DbFortBuild> userForts = await fortRepository.Builds.Where(x =>
            x.ViewerId == viewerId
        ).ToListAsync();
```

**Comment by SapiensAnatis** [comment]
> Lots of changes with this last commit. There should be way less database lookups. And I ran CSharpier this time!

Glad to hear it. I think you are getting there but perhaps the `IQueryable` stuff has tripped you up a little. It's looking a lot better though.

File: DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs
Comment (jaredstrong89) [reply DragaliaAPI/DragaliaAPI/Features/StorySkip/StorySkipService.cs]: I did not realize this. I changed these objects to Lists.

**Review by SapiensAnatis — APPROVED** [review]
I've done some testing of this locally and it appears to work pretty well! Happy with the code as well.

I'll merge it in tomorrow after work just so I'm on hand in case any issues arise. Thanks once again for working on this.

**Pull request merged**
